unified_planning/engines/sequential_simulator.py

```diff
@@ -207,10 +183,52 @@ def _get_applicable_events(self, state: "up.model.ROState") -> Iterator["Event"]
         :param state: The state where the formulas are evaluated.
         :return: an Iterator of applicable Events.
         """
-        for events in self._events.values():
-            for event in events:
-                if self.is_applicable(event, state):
-                    yield event
+        # if the problem was never fully grounded before,
+        # ground it and save all the new events. For every event
+        # that is applicable, yield it.
+        # Otherwise just return all the applicable events
+        if not self._all_events_grounded:
+
+            # total grounding
+            self._all_events_grounded = True
+            grounding_result = self._grounder.ground_problem()
+            grounded_problem: "up.model.Problem" = cast(
+                up.model.Problem, grounding_result.problem
+            )
+            lift_map = grounding_result.map_back_action_instance
+            assert lift_map is not None
+
+            # for every grounded action, translate it in an Event
```

Comment:
As you already recognized in the comment, this code does something reusable, make a function out of it! :smile: